src/yaml-support/yaml-schema.ts

```diff
@@ -0,0 +1,165 @@
+                        apiVersion,
+                        name,
+                        kind,
+                        ...node
+                    };
+
+
+                });
+            }
+        } else {
+            _definitions[name.toLowerCase()] = {
+                name,
+                ...node
+            };
+        }
+
+        toFixed.push(node['properties']);
+    }
+
+    for (let fix of toFixed ) {
+        patchOnRef(fix);
+    }
+}
+
+export async function registerYamlSchemaSupport() {
+    loadSchema(KUBERNETES_SCHEMA);
+    let yamlPlugin;
+    const installedExtensions: any[] = vscode.extensions.all;
+    for (let ext of installedExtensions) {
+        if (ext.id === VSCODE_YAML) {
+            try {
+                yamlPlugin = await Promise.resolve(ext.activate());
+                break;
+            } catch (error) {
+                console.log('Failed to activate the Yaml Extension: ' + error);
+            }
+            break;
+        }
+    }
+    if (!yamlPlugin) {
+        console.log('vscode-yaml plugin doesn\'t expose the entry to register custom schema provider.');
+        return;
+    }
+    // register for kubernetes schema provider
+    yamlPlugin.registerContributor('kubernetes', (resource) => {
+        const textEditor = vscode.window.visibleTextEditors.find((editor) => editor.document.uri.toString() === resource);
+        if (textEditor) {
+            const yamlDocs = yamlLocator.getYamlDocuments(textEditor.document);
+            let choices: string[] = [];
+            yamlDocs.forEach((doc)=> {
+                // if the yaml document contains apiVersion and kind node, it will report it is a kubernetes yaml
+                // file
+                const rootMapping = doc.nodes.find((node) => node.kind === 'MAPPING');
+                // TODO, unwrap quotes
+                const apiVersion = rootMapping.mappings.find((map) => map.key && map.key.raw === 'apiVersion');
+                const kind = rootMapping.mappings.find((map) => map.key && map.key.raw === 'kind');
+                if (apiVersion && kind) {
+                    choices.push(apiVersion.value.raw + GROUP_VERSION_KIND_SEPARATOR + kind.value.raw);
+                }
+            });
+            choices = _.uniq(choices);
+            if (choices.length > 1) {
+                return 'kubernetes://combine/' + _.uniq(choices).join('+');
+            }
+            if (choices.length === 1) {
+                return 'kubernetes://manifest/'  + choices[0];
+            }
+            return undefined;
+        }
+    }, (uri) => {
+        let _uri = Uri.parse(uri);
+        if (!_uri.path || ! _uri.path.startsWith('/')) {
+            return '';
+        }
+        if (_uri.path.includes('+')) {
+            const manifestRef = _uri.path.slice(1).split('+').map((r) => {
+                return {$ref: `kubernetes://manifest/${r}`};
```

Comment:
You're clearly lining up URIs between here and the `resource` lambda so pull out the URI generation into a shared method.